lib/compiler_rt/x86_cpu_model.zig

```diff
@@ -0,0 +1,165 @@
+            if (comptime !@hasField(Feature, f.name)) continue;
+            if (detected_features.isEnabled(index)) setFeature(&features, @field(Feature, f.name));
+        }
+
+        const isDetected = struct {
+            fn isDetected(set: Target.Cpu.Feature.Set, feats: []const Target.x86.Feature) bool {
+                for (feats) |f| {
+                    if (!set.isEnabled(@intFromEnum(f))) return false;
+                }
+                return true;
+            }
+        }.isDetected;
+
+        // Compared to upstream LLVM,
+        //  - FEATURE_LM -> @"64bit"
+        if (isDetected(detected_features, &.{ .@"64bit", .sse2 })) {
+            setFeature(&features, .x86_64);
+
+            // Compared to upstream LLVM,
+            //  - FEATURE_CMPXCHG16B -> cx16
+            //  - FEATURE_LAHF_LM -> sahf
+            if (isDetected(detected_features, &.{ .cx16, .popcnt, .sahf, .sse4_2 })) {
+                setFeature(&features, .x86_64_v2);
+
+                if (isDetected(detected_features, &.{ .avx2, .bmi, .bmi2, .f16c, .fma, .lzcnt, .movbe })) {
+                    setFeature(&features, .x86_64_v3);
+
+                    if (isDetected(detected_features, &.{ .avx512bw, .avx512cd, .avx512dq, .avx512vl })) {
+                        setFeature(&features, .x86_64_v4);
+                    }
+                }
+            }
+        }
+
+        break :blk features;
+    };
+
+    cpu = .{
+        .vendor = detected.vendor,
+        .type = detected.type,
+        .subtype = detected.subtype,
+        .features = feature_set[0..1].*,
+    };
+    cpu_extra_features = feature_set[1..].*;
+
+    return 0;
+}
+
+const Model = extern struct {
+    vendor: Vendor = .unknown,
+    type: Type = .unknown,
+    subtype: Subtype = .unknown,
+    features: [1]u32 = .{0},
+};
+
+const FeatureSet = [feature_set_size]u32;
+comptime {
+    for (@typeInfo(Feature).@"enum".fields) |f| {
+        if (f.value >= @bitSizeOf(FeatureSet))
+            @compileError(@import("std").fmt.comptimePrint("Feature.{s} ({}) bitindex too large", .{ f.name, f.value }));
+    }
+}
+
+fn hasFeature(set: FeatureSet, f: Feature) bool {
+    const f_value = @intFromEnum(f);
+    const bit = set[f_value / 32] & (@as(u32, 1) << @as(u5, @intCast(f_value % 32)));
+    return bit != 0;
+}
+fn setFeature(set: *FeatureSet, f: Feature) void {
+    const f_value = @intFromEnum(f);
+    set[f_value / 32] |= (@as(u32, 1) << @as(u5, @intCast(f_value % 32)));
+}
+
+const Vendor = x86.enums.Vendor;
+const Type = x86.enums.Type;
+const Subtype = x86.enums.Subtype;
+const Feature = x86.enums.Feature;
+
+const feature_set_size = blk: {
+    var max_index: comptime_int = 0;
+    for (@typeInfo(Feature).@"enum".fields) |f| {
+        if (f.value > max_index) max_index = f.value;
+    }
+    break :blk (max_index + 32) / 32;
+};
+
+comptime {
+    // Sanity checking that the feature sets in std.Target.x86 and the autogenerated
+    // ones in @This().Feature are compatible.
```

Comment:
I don't love the idea of doing this here due to compile times. Since we will be running the new script during [the LLVM upgrade process](https://github.com/ziglang/zig/wiki/LLVM-Upgrade-Process) much like `update_cpu_features.zig`, I think it would make more sense to just do these sanity checks in the script and require that it's built with a compiler that has an up-to-date `std.Target.x86`.

Reply:
You might even consider merging the new script into `update_cpu_features.zig` since they should both be run together anyway.